### PR TITLE
GUIFormspecMenu: Add animated_image[] element 

### PR DESCRIPTION
**Pull request by ghost**
This is a rebased/updated version of #9032, intended to address the issues that were brought up during review. Since @kilbith wasn't in a position to rebase and complete the original PR, I am adopting in hopes of getting this past the home stretch!

A word of caution. The work on this was split across about a month (with a ~3-week gap) due to some irl issues, so while the main functionality works I can't guarantee that everything was properly handled. *Please test/review carefully!*

I should also point out that I haven't squashed the original PR's changes yet, I wanted the separation to be clear so that folks could more easily see what I've changed.

## How to Test
To test this PR, create a formspec using the new `animated_image` element and ensure that it renders as expected.

For your convenience, I've attached a minimal test mod: \
[test_animimage.zip](https://github.com/minetest/minetest/files/4145293/test_animimage.zip) This mod adds the `/anim` chat command, which displays a fomspec with an animated image inside. I recommend making further changes to the formspec for testing, but this should simplify the process of setting up your test environment.

### CODE REVIEW

File: src/client/texture.cpp
Comment (DS-Minetest) [inline src/client/texture.cpp]: This cast won't work correctly for `m_frame_count` greater than 9. I suggest using something like [`std::to_string` ](https://en.cppreference.com/w/cpp/string/basic_string/to_string) instead:
```suggestion
		m_frame_names[i] = base_name + "_" + std::to_string(i + 1) + ".png";
```

File: src/client/texture.cpp
Comment (DS-Minetest) [inline src/client/texture.cpp]: Shouldn't this be `m_frame_time %= m_frame_duration * m_frame_count;`?

File: src/client/texture_pool.h
Comment (DS-Minetest) [inline src/client/texture_pool.h]: What do you need `<iostream>` for?
```suggestion
```

File: src/gui/guiAnimatedImage.cpp
Comment (DS-Minetest) [inline src/gui/guiAnimatedImage.cpp]: Error message spam in draw functions can be problematic.

File: src/client/texture.cpp
Comment (Df458) [reply src/client/texture.cpp]: No, it's correct. `m_frame_time` only refers to the current frame here, it's used to advance `m_frame_idx` at an interval of `m_frame_duration` ms.

If we change the line to `m_frame_time %= m_frame_duration * m_frame_count;`, what happens is that the value gets high enough to advance `m_frame_idx` up to `m_frame_count` times per-frame which messes up the playback.

File: src/gui/guiAnimatedImage.cpp
Comment (Df458) [reply src/gui/guiAnimatedImage.cpp]: I just ported that log over from the original code. Would you prefer I remove it altogether, or should it be moved somewhere specific?

File: src/client/texture.cpp
Comment (Df458) [reply src/client/texture.cpp]: Good catch! I've applied the fix and updated the test mod to use 10+ frames to confirm that it works.

File: src/gui/guiAnimatedImage.cpp
Comment (DS-Minetest) [inline src/gui/guiAnimatedImage.cpp]: Why shouldn't the new element be clippable?
```suggestion
				NoClip ? nullptr : &AbsoluteClippingRect, colors, true);
```

File: src/gui/guiAnimatedImage.cpp
Comment (DS-Minetest) [reply src/gui/guiAnimatedImage.cpp]: I don't know. Finding this error elsewhere (eg, in the parser) will probably be difficult.

Comment (Df458) [reply src/gui/guiAnimatedImage.cpp]: Actually... when the file isn't there it will normally generate a replacement and log an error anyway. I've decided to remove it, but I'll leave this unresolved in case others disagree.

File: src/gui/guiFormSpecMenu.cpp
Comment (DS-Minetest) [inline src/gui/guiFormSpecMenu.cpp]: (`{` should be on the same line as the `if` parenthesis (`if (blabla\nbla) {`.)

Comment (DS-Minetest) [inline src/gui/guiFormSpecMenu.cpp]:
```cpp
	if (parts.size() != 3 &&
			(parts.size() < 3 || m_formspec_version <= FORMSPEC_API_VERSION)) {
		errorstream << "Invalid animated image element(" << parts.size() << "): '" << element << "'"  << std::endl;
		return;
	}
...
```

File: src/gui/guiFormSpecMenu.cpp
Comment (DS-Minetest) [inline src/gui/guiFormSpecMenu.cpp]: I'd not support the old image pos and geom format or what that is.

File: src/gui/guiFormSpecMenu.cpp
Comment (DS-Minetest) [inline src/gui/guiFormSpecMenu.cpp]:
```suggestion
			warningstream << "invalid use of animated image without a size[] element" << std::endl;
```

File: src/gui/guiFormSpecMenu.cpp
Comment (DS-Minetest) [inline src/gui/guiFormSpecMenu.cpp]: Clipping should be allowed imo.
```suggestion
```

File: src/gui/guiFormSpecMenu.cpp
Comment (DS-Minetest) [inline src/gui/guiFormSpecMenu.cpp]: `e` has to be dropped somewhere (eg. at the end of this function).

File: src/gui/guiFormSpecMenu.cpp
Comment (DS-Minetest) [inline src/gui/guiFormSpecMenu.cpp]: This should probably be done a bit earlier, before anything is drawn, eg. in line 3316.

File: src/gui/guiFormSpecMenu.h
Comment (DS-Minetest) [inline src/gui/guiFormSpecMenu.h]: Move the include into the .cpp file and add a `class TexturePool;` to the header (below this comment).

File: src/gui/guiAnimatedImage.cpp
Comment (Df458) [reply src/gui/guiAnimatedImage.cpp]: Interestingly, `image[]` doesn't appear to be clippable either. I'll make a PR for that while I'm at it.

**Comment by DS-Minetest** [comment]
The new files could use some code comments.

File: src/gui/guiFormSpecMenu.cpp
Comment (Df458) [reply src/gui/guiFormSpecMenu.cpp]: I like that, but it doesn't follow the existing convention used in every other `parseX()` function, as far as I can tell. I'll make the change if another reviewer agrees with this.

File: src/gui/guiFormSpecMenu.cpp
Comment (DS-Minetest) [inline src/gui/guiFormSpecMenu.cpp]: You have to delete this in the destructor (line 148).

Comment (Df458) [reply src/gui/guiFormSpecMenu.cpp]: Whoops, you're right. I need to be more careful, been making some really dumb errors on this PR...

File: src/gui/guiFormSpecMenu.cpp
Comment (Df458) [reply src/gui/guiFormSpecMenu.cpp]: So just to confirm, you're suggesting that we just always use real coordinates since this is a new control?

Comment (DS-Minetest) [reply src/gui/guiFormSpecMenu.cpp]: Yes, I think so.

Comment (Df458) [reply src/gui/guiFormSpecMenu.cpp]: The result looks weird, but the old non-real coordinates result also looks weird and real coordinates are preferable anyway so I've pushed the change.

File: src/client/texture_pool.cpp
Comment (DS-Minetest) [inline src/client/texture_pool.cpp]: Could you maybe instead use a pointer for `texture_idx` instead of a reference? That would make it clearer at the function call that it will be overwritten by the function.
The code style guidelines also say:
> Don't pass non-const references to functions. 

https://dev.minetest.net/Code_style_guidelines#Do_not_be_too_C.2B.2By

File: src/gui/guiAnimatedImage.cpp
Comment (Df458) [reply src/gui/guiAnimatedImage.cpp]: Since an error is still logged with the logs removed altogether and no-one is piping up about this, I'm marking this as resolved.

File: src/gui/guiFormSpecMenu.cpp
Comment (Df458) [reply src/gui/guiFormSpecMenu.cpp]: Actually, I'm just going to be the change I want to see in the world and do it. If anyone disagrees in later reviews, I can always revert the commit.

**Comment by Df458** [comment]
There hasn't been any activity in about a week, but aside from squashing this PR has been all set for review/approval/merge for a while. Any maintainers have the time to look into it this week?

**Comment by v-rob** [comment]
Could support for other file formats like `.jpg` be added by splitting the filename at the last period and then concatenating the file number between the name and the extension?

**Comment by Df458** [comment]
> Could support for other file formats like .jpg be added by splitting the filename at the last period and then concatenating the file number between the name and the extension?

Sounds reasonable. I'll look into that tomorrow. Another possibility would be to have some sort of token representing the number in the filename, but I'm not sure it's worth the effort to do. It could potentially make the API clearer though... any opinions on this?

**Comment by v-rob** [comment]
I don't think having symbols is necessary; it'll just complicate everything.  If we leave the numbers as-is, it will be a simple naming convention rule that everyone will consistently follow.  But having support for multiple filenames is a little more of important functionality, even if most people use PNG anyway.

**Comment by Df458** [comment]
I've added support for supplying an extension and rebased. For simplicity, I'm currently assuming `.png` if the user doesn't supply an extension since I'm fairly certain that anything without an extension will fail to load since the extension seems to matter to the loader (indeed, `.jpeg` appears to fail loading despite support for `.jpg`, I believe the bug is unrelated to this feature though).

**Comment by ClobberXD** [comment]
Tested using the test mod provided in the first post, and also tried different combinations of frame duration and frame count. PR works as expected :+1: 

I thought the duration of jpg images were inconsistent, until I realised that test_1.jpg and test_6.jpg were the same image, lol :)

**Comment by kilbith** [comment]
@SmallJoker [Rebase needed] can be removed.

Other plans I had in the original PR:
- Add `animated_image_button[]`
- Add a `on_hover` spec (bool - default to false), so the image animates when hovered.

**Comment by SmallJoker** [comment]
@kilbith What was the idea of using `texture_name_%i.png`? Sounds are named `sound_name.%i.ogg` and animated frames in the node definition is a single texture, but with all frames stacked along the Y axis.
I'm asking because it could be helpful to adapt one practice from there rather than introducing a new mechanism. The latter (stacked frames) would allow re-using node definition textures and improve the media loading & saving process due to less files, but more data.

**Comment by kilbith** [comment]
@SmallJoker IMO stacked frames fit to a tile because the display's width is guaranteed to be the same. For the GUI or anything else, it's a different story - you may have various textures of very different sizes, and it's not easy to stack them when the gap is too important. Moreover, one could extract a video frame by frame with `ffmpeg`, and play this "video" in a formspec.

EDIT: This PR should adopt the `filename.%i.extension` convention.

**Comment by Df458** [comment]
I've updated the code to use the convention mentioned above, and updated the test mod to do the same. As for the other comments:

- I can appreciate the benefits of a button with an animated image, though at this point it might make sense to add support for buttons with custom content instead (I expect this would be fairly straightforward thanks to #9090). Such a feature would fill the need, and open the door to further possibilities at the same time.
- Having inherited this PR rather than starting it myself, I have no strong opinion re: stacked frames. I'm a little worried about the potential for further prolonging a PR that's been mostly done for quite some time, but I can see the argument for consistency there as well. I'll leave the final decision to the core devs, and make changes if necessary.

**Comment by Df458** [comment]
Since no-one has come in with further opinions and git makes backing out of changes trivial, I've taken the liberty of changing this PR to use stacked frames instead of separate files. Hopefully this will let the conversation continue.

@SmallJoker feel free to take another look and give your take on this.

File: src/gui/guiAnimatedImage.cpp
Comment (SmallJoker) [inline src/gui/guiAnimatedImage.cpp]: Please use tabs.

File: src/gui/guiAnimatedImage.cpp
Comment (SmallJoker) [inline src/gui/guiAnimatedImage.cpp]: I'd output an error here. It's clearly a wrong format for this formspec element.

File: src/gui/guiAnimatedImage.cpp
Comment (SmallJoker) [inline src/gui/guiAnimatedImage.cpp]: Tabs again

File: src/gui/guiAnimatedImage.cpp
Comment (SmallJoker) [inline src/gui/guiAnimatedImage.cpp]: Getting rid of `step_duration` 
```C++
u64 new_global_time = porting::getTimeMs();
// Determine the delta time to step
if (m_global_time > 0)
	m_frame_time += new_global_time - m_global_time;
m_global_time = new_global_time;
```

File: src/gui/guiFormSpecMenu.cpp
Comment (SmallJoker) [inline src/gui/guiFormSpecMenu.cpp]: Line is longer than 80 characters. Please break it.

File: src/gui/guiFormSpecMenu.cpp
Comment (SmallJoker) [inline src/gui/guiFormSpecMenu.cpp]: This would look better IMO:
```C++
FieldSpec spec(
	"",
	L"",
	L"",
	258 + m_fields.size()
);
```

File: src/gui/guiFormSpecMenu.h
Comment (SmallJoker) [inline src/gui/guiFormSpecMenu.h]: Old code does not always follow the coding style guidelines. Please move the pointer placement towards the variable.

**Review by SmallJoker — COMMENTED** [review]
The code is slim and works as expected. It needs a few stylistic changes, though.

**Comment by SmallJoker** [comment]
A second core dev opinion would indeed be better than to rely on my comments only. Now that you've changed the PR there are no concept objections I could possibly have.

File: src/gui/guiFormSpecMenu.cpp
Comment (Df458) [reply src/gui/guiFormSpecMenu.cpp]: You're right that the line is too long, and I've gone in and broken it. I think the result is harder to read though, I'd like your opinion on how I did the break and if you see a better spot to do it.

File: src/gui/guiAnimatedImage.cpp
Comment (Df458) [reply src/gui/guiAnimatedImage.cpp]: Done, I think the message is good but feel free to re-check now!

Comment (SmallJoker) [reply src/gui/guiAnimatedImage.cpp]: Good so far, but the line needs breaking (>80 characters). It also wouldn't hurt to include a hint about the error source, example: `animated_image[]: Invalid texture ...`. I know the element name does not fit into this class but it'll be helpful to modders.

**Review by SmallJoker — APPROVED** [review]

File: src/gui/guiAnimatedImage.cpp
Comment (Df458) [reply src/gui/guiAnimatedImage.cpp]: Looks like you went and fixed it for me, thank you!
Does that mean this conversation can be resolved?

**Comment by kilbith** [comment]
I think it's a very bad idea to force using real coordinates implicitly. I do have some complex formspecs which are hard to transit to real coordinates, so they keep using the old system, but I still want to pick up animated images for them.

And it was the same mistake for `hypertext`.

File: src/gui/guiAnimatedImage.cpp
Comment (SmallJoker) [reply src/gui/guiAnimatedImage.cpp]: Yes. If requested changes were addressed feel free to mark my comments as resolved.

File: src/gui/guiFormSpecMenu.cpp
Comment (SmallJoker) [reply src/gui/guiFormSpecMenu.cpp]: For two lines it's fine to break it like this. If it were longer, passing the template to `snprintf` would be a good solution.
Optimally binary operators would be placed to the end of the previous line. I think both is fine, so I'm not going to enforce that.

**Comment by Df458** [comment]
There was support previously, but it @DS-Minetest suggested that I remove it. I could potentially reinstate support, but I'd rather stick to the current policy...

@SmallJoker, I believe you were involved in the approval process for the hypertext PR. What should be done with non-`real_coordinates` mode on new elements? Same as there, or was that an oversight?

**Comment by SmallJoker** [comment]
> or was that an oversight?

That, and here the same thing happened. Copying the scaling code from the `image[]` element should be possible and provide "good enough" results. Prior `real_coordinates` the scaling and positioning is broken anyway, so I wouldn't spend too much effort into making new elements backwards compatible.

**Comment by kilbith** [comment]
So I have to spend hours to convert my formspec to real coordinates in order to use `animated_image` or `hypertext`? How is that hurting to add a one-liner to support the old coordinates system? And how is that clear to someone that these elements works only correctly with real coordinates, since it's implicit?

Are you stupid or what?

**Comment by Df458** [comment]
> Copying the scaling code from the image[] element should be possible and provide "good enough" results.

That's more or less what was there originally. Since we've confirmed that the original change wasn't deliberate, interest has been shown, and the cost is negligible, I've re-instated the feature. @SmallJoker  let me know if that's a deal-breaker and I can back out, but since it's so cheap to support I don't really mind keeping it.

Slightly off-topic: @kilbith, I understand that getting things right is important to you and I'm legitimately glad that you're advocating for it... but I feel that your comments sometime cross the line from spirited debate into creating unnecessary drama. It doesn't bother me personally, but I get the feeling that this sort of thing will be a source of issues sooner or later.

File: src/gui/guiFormSpecMenu.cpp
Comment (SmallJoker) [inline src/gui/guiFormSpecMenu.cpp]: Please use tabulators.

Comment (Df458) [reply src/gui/guiFormSpecMenu.cpp]: Sorry for screwing that up a second time...I've taken measures to ensure that it won't happen again.

**Comment by SmallJoker** [comment]
:+1: PR looks good again.

File: src/gui/guiAnimatedImage.cpp
Comment (rubenwardy) [inline src/gui/guiAnimatedImage.cpp]: Why isn't colon_position checked?

```suggestion
	if (comma_position != std::string::npos && colon_position != std::string::npos &&
			comma_position > colon_position) {
```

File: src/gui/guiAnimatedImage.cpp
Comment (rubenwardy) [inline src/gui/guiAnimatedImage.cpp]: this can be initialised in the definition

File: src/gui/guiAnimatedImage.cpp
Comment (rubenwardy) [inline src/gui/guiAnimatedImage.cpp]: duplicated initialisation

File: src/gui/guiFormSpecMenu.cpp
Comment (rubenwardy) [inline src/gui/guiFormSpecMenu.cpp]: Please don't De Morgan this - I would prefer the more obvious condition and the one that is throughout this fileused 

```suggestion
			!(parts.size() > 3 && m_formspec_version > FORMSPEC_API_VERSION)) {
```

Comment (SmallJoker) [reply src/gui/guiFormSpecMenu.cpp]: The current logic also shows that `parts.size() != 3 &&` is redundant.

Comment (Df458) [reply src/gui/guiFormSpecMenu.cpp]: Fine by me, that's more or less what I originally had when I submitted the PR.

File: src/gui/guiAnimatedImage.cpp
Comment (Df458) [reply src/gui/guiAnimatedImage.cpp]: It is...just in a dangerously indirect fashion. Good catch.

File: src/gui/guiFormSpecMenu.cpp
Comment (rubenwardy) [reply src/gui/guiFormSpecMenu.cpp]: Were you asked to do this by someone?

Comment (Df458) [reply src/gui/guiFormSpecMenu.cpp]: Sort of? It's in one of the reviews by @DS-Minetest. The main aspect of the review was to reduce nesting though, so this particular mistake is still on me.

**Review by rubenwardy — APPROVED** [review]

**Pull request merged**
